coderd/prometheusmetrics/aggregator.go

```diff
@@ -64,15 +68,58 @@ type annotatedMetric struct {
 	expiryDate time.Time
 }
 
+type metricKey struct {
+	username      string
+	workspaceName string
+	agentName     string
+	templateName  string
+
+	metricName string
+	labelsStr  string
+}
+
+func (m1 metricKey) Equal(m2 metricKey) bool {
+	return m1.metricName == m2.metricName &&
+		m1.labelsStr == m2.labelsStr &&
+		m1.username == m2.username &&
+		m1.workspaceName == m2.workspaceName &&
+		m1.agentName == m2.agentName &&
+		m1.templateName == m2.templateName
+}
+
+func hashKey(req *updateRequest, m *agentproto.Stats_Metric) metricKey {
+	var sb strings.Builder
+	for _, label := range m.GetLabels() {
+		if label.Value == "" {
+			continue
+		}
+
+		_, _ = sb.WriteString(label.Name)
+		_ = sb.WriteByte('=')
+		_, _ = sb.WriteString(MetricLabelValueEncoder.Replace(label.Value))
+		_ = sb.WriteByte(',')
+	}
+	labels := strings.TrimRight(sb.String(), ",")
+
+	return metricKey{
+		username:      req.username,
+		workspaceName: req.workspaceName,
+		agentName:     req.agentName,
+		templateName:  req.templateName,
+		metricName:    m.Name,
+		labelsStr:     labels,
+	}
+}
+
 var _ prometheus.Collector = new(MetricsAggregator)
 
 func (am *annotatedMetric) is(req updateRequest, m *agentproto.Stats_Metric) bool {
-	return am.username == req.username &&
+	return am.Name == m.Name &&
```

Comment:
I think `is()` is now only called from tests, so it can be removed and the tests adapted to check equality of the hash keys.

Reply:
Removed `is()` and the whole set of internal tests.